initial stab at process.cpuUsage()

pmuellr opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `make -j8 test` (UNIX) or `vcbuild test nosign` (Windows) pass with
  this change (including linting)?
- [ ] Is the commit message formatted according to [CONTRIBUTING.md][0]?
- [ ] If this change fixes a bug (or a performance problem), is a regression
  test (or a benchmark) included?
- [ ] Is a documentation update included (if this change modifies
  existing APIs, or introduces new ones)?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Affected core subsystem(s)

process

[0]: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-3-commit

### Description of change

Adds the function `cpuUsage()` to the global `process` object.  Proposed doc is in the PR, and immediate below:

**process.cpuUsage()**

Returns the user and system cpu time usage of the current process, in an object with properties `user` and `system`, whose values are arrays of integers as `[seconds, microseconds]` tuples. These values measure time spent in user and system code respectively, and may end up being greater than actual elapsed time if multiple cpu cores are performing work for this process.

```js
console.log(process.cpuUsage());
// { user: [ 0, 43979 ], system: [ 0, 12423 ] }

setTimeout(() => {
  var startTime = process.hrtime();

  // spin the CPU for 500 milliseconds
  while (true) {
    var time = process.hrtime(startTime);

    // calculate elapsed wall clock time in millis
    var elapsed = time[0] * 1000 + time[1] / 1000 / 1000;

    if (elapsed > 500) break;
  }

  console.log(process.cpuUsage());
  // { user: [ 0, 554966 ], system: [ 0, 14881 ] }
}, 1000);
```

mscdex commented on src/node.cc:
```diff
@@ -2137,6 +2137,32 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+  Local<ArrayBuffer> ab = args[0].As<Uint32Array>()->Buffer();
+  uint32_t* fields = static_cast<uint32_t*>(ab->GetContents().Data());
+
+  // results passed as split values, reassembed in JS
```
s/reassembed/reassembled/

mscdex commented on src/node.js:
```diff
@@ -208,6 +208,25 @@
+    const _cpuUsage = process.cpuUsage;
+    const cpuValues = new Uint32Array(8);
+
+    process.cpuUsage = function cpuUsage(ar) {
```
What is the `ar` parameter for?

pmuellr replied:
Woops, leftover garbage from copying `hrtime()` bits, above this.  Should **not** take any arguments (`ar` should be removed).

Trott commented:
Should get at least a minimal test that confirms the object returned by `process.cpuUsage()` conforms with what is documented.

pmuellr commented:
> Should get at least a minimal test that confirms the object returned by process.cpuUsage() conforms with what is documented.

absolutely @Trott 

Figured I'd do the minimal work to see if there's even interest in this, if the approach is generally correct, etc.

The most basic test is to ensure that subsequent values are monotonically increasing, or the same.  Would be nice to have one that slept for a 100 ms or so, and then made sure the old / new values were within some range of reasonable-ness, but I fear could easily become flaky.  I will also peek at libuv's tests for this.

BTW, purty sure this my first Node.js PR!

evanlucas commented:
I really like the idea, but should we be adding this onto `process`? I think we should try to avoid adding anything to `process` unless absolutely necessary. Maybe a metrics builtin or something like that?

YuriSolovyov commented:
Maybe put it to `os` module?

mscdex commented:
The `os` module is typically for more global stuff. I think `process` is a good fit for it, whether it should go directly on it or whether it and `memoryUsage()` would go on a new `.metrics`-like namespace like @evanlucas suggested.

ronkorving commented on src/node.cc:
```diff
@@ -2137,6 +2137,32 @@ void Hrtime(const FunctionCallbackInfo<Value>& args) {
+// to access ru_utime (user CPU time used) and ru_stime (system CPU time used),
+// which are uv_timeval_t structs (long tv_sec, long tv_usec) as tuples
+// similar to process.hrtime() values.  Only it's seconds/microseconds instead
+// of hrtime()'s seconds/nanoseconds.
```
To avoid confusion in user land, would it make sense to turn them into nanoseconds?

pmuellr replied:
Have to admit, I'm a bit torn on this:

- on one hand, `rusage()` returns seconds / microseconds, and so I'd kinda like to match that
- on the other hand, `process.hrtime()` returns [seconds, nanoseconds], so would be nice to be consistent

I'm kinda leaning towards your suggestion, returning [seconds, nanoseconds], especially since folks may be using both `hrtime()` and `cpuUsage()` together, to get a "cpu usage %" type value.

If we want to do that, then I think we'd just doc that the resolution is currently at the microsecond level, but the value returned is nanoseconds.

pmuellr replied:
If we align more with `hrtime()`, perhaps it also makes sense to provide a "diff" function, like `hrtime()`, where if you pass a previous result of `cpuUsage()` into the function, it will return a "diff" of the times, instead of the current time.  For one of the primary use cases I'm considering, I'd be wanting a diff anyway, so this cuts down a little bit on the amount of code I'd be writing.

ronkorving replied:
That's interesting. Would love to see some feedback from others on this.

eljefedelrodeodeljefe replied:
generally +1. I'd favor nanoseconds for consistency also. Also +1 on the diffing. Would be nicely consistent to vanilla js also. 

Just a quick thought: Maybe it's worth taking this even further, whether `let p = process.cpuUsage(); p.diff(); p.diff()` might be a thing.

jasnell commented:
Needs some form of test case but generally LGTM

Fishrock123 commented:
@pmuellr Could you go over the checklist in the OP? Thanks! :D

pmuellr commented:
@Fishrock123 sure; let me refresh the code given the changes mentioned here (still in `process`, but doing the diff, and using nanos instead of micros), add some tests, and I'll ensure the checklist is ready to go.  Prolly end of this week.

bnoordhuis commented:
@pmuellr Ping.

pmuellr commented:
Hoping to look at this next week.  No prob if someone wants to help out tho!

On Wed, Mar 23, 2016 at 2:19 PM, Ben Noordhuis <notifications@github.com>
wrote:

> @pmuellr <https://github.com/pmuellr> Ping.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/5565#issuecomment-200475882>
>



-- 
Patrick Mueller
http://muellerware.org

eljefedelrodeodeljefe commented:
I wanted to take this up. Having had a look though, I was thinking whether it wouldn't be better to expose the `uv_getrusage()` as a whole. 
Most of the values it returns might not be useful in obvious ways and currently NT would only support `utime` and `stime`, but this way we would have a less opinionated feature.

Any thoughts?

pmuellr commented:
@eljefedelrodeodeljefe I often think that as well; let me write my C code in JS!

OTOH, given the current thought on aligning w/ `hrtime()` on nanoseconds vs microseconds, and allowing the diff, makes that thought less interesting.

No reason we can't have two APIs tho!

eljefedelrodeodeljefe commented:
Sure. Then let's proceed with it as is, when you have time.

pmuellr commented:
Think everything's pretty much done here.  Latest commit a few minutes ago: 

* changed resolution to nanoseconds instead of micros, to
  align with process.hrtime()
* take a previous result as an arg and return a diff, like
  process.hrtime()
* fixed doc
* added tests
commit c23841e4c737cb43807e6ddcf7a08a9080178dde

Still need to rebase onto master, generate a final commit message.

eljefedelrodeodeljefe commented:
Looking good. LGTM from here.

pmuellr commented:
Looks like I picked a bad time to work on this; shoulda gotten the patch in before the node.js refactor, or started after.  Bit beyond me how to rebase w/ master in this branch, given the refactor.  I may end up building a new branch based on master ...

jasnell commented:
The changes to src/node.js can be moved into a separate module inside /lib/internal/process, similar to how I did with the process warnings. Then, inside lib/bootstrap_node.js you can require that following the same pattern. 
/cc @Fishrock123

Fishrock123 commented:
This should probably be in `lib/internal/process.js`.

As for rebasing, I suggest just moving the existing changes out of node.js into a new file for storage, rebasing, and re-pasting them into the new file. Everything else should work as-is, I think.

pmuellr commented:
The current state is that I refactored the init code so that it's done in the same way has `hrtime()`.  Test still work, so ... \o/.

Last thing to do is the commit message - guess I'll troll the existing ones and look for a similar one to copy from, but link would be great.

But this PR has gone haywire, it appears, in GH.  Looks ok in my local repo.  Not sure what's going on, but almost assuredly it's confused by cross-repo, squashed commits force pushed, rebased from  `nodejs:v5.x` to master.  That's a lot to deal with!

<img width="636" alt="initial_stab_at_process_cpuusage___by_pmuellr_ _pull_request__5565_ _nodejs_node" src="https://cloud.githubusercontent.com/assets/25117/14282326/e127cfa2-fb0b-11e5-8862-858b6f1cee95.png">

Seems easiest if I close this, reopen a new one based on master; reference this for the existing comment thread.

Or other tips?

pmuellr commented:
ya, this looks a lot cleaner :-)

* https://github.com/nodejs/node/compare/master...pmuellr:cpuUsage2?expand=1

eljefedelrodeodeljefe commented:
Now you dragged in all the commits since ad0ce574320966ab2c7410a2c2ec8530e2bae500 . Think this needs another run :( Since the changes are not that big I suggest the good old `git branch -m pmuellr/cpuUsage pmuellr/cpuUsage_old` and checkout out a new branch from a fresh master.

pmuellr commented:
My git foo is weak.  The rebases I've done on my clone/fork look ok, but in the PR, they're pulling in stuff from the old branch I started on (some 5.x branch, which was incorrect, shouda been master from the start).  

So, giving up on this branch / PR.  

I've got a clean branch / PR here as a replacement: https://github.com/nodejs/node/pull/6157
